externallib.php

```diff
@@ -0,0 +1,688 @@
+<?php
+require_once("$CFG->libdir/externallib.php");
+class moodle_gradereport_quizanalytics_external extends external_api
+{
+    public static function quizanalytics_analytic_parameters()
+    {
+        return new external_function_parameters(
+            array(
+                'quizid' => new external_value(PARAM_INT, 'quiz id')
+            )
+        );
+    }
+    public static function quizanalytics_analytic($quizid)
+    {
+        global $DB, $USER, $PAGE, $CFG;
+        $quiz = $DB->get_record('quiz', array('id' => $quizid));
+
+        $attemptssql = "SELECT * FROM {quiz_attempts}
+          WHERE state = 'finished' AND sumgrades IS NOT NULL AND quiz = ?";
+
+        $totalquizattempted = $DB->get_records_sql($attemptssql, array($quizid));
+
+        $usersgradedattempts = $DB->get_records_sql($attemptssql . " AND userid = ?", array($quizid, $USER->id));
+
+        $totalnoofquestion = $DB->get_record_sql("SELECT COUNT(qs.id) as qnum
+                      FROM {quiz_slots} qs, {question} q WHERE q.id = qs.id
+                      AND qs.quizid = ? AND q.qtype != ?", array($quizid, 'description'));
+        if (!empty($usersgradedattempts)) {
+            /**
+             * Return the part of random color.
+             */
+            function random_color_part()
+            {
+                return str_pad(dechex(mt_rand(0, 255)), 2, '0', STR_PAD_LEFT);
+            }
+            /**
+             * Return the random color.
+             */
+            function random_color()
+            {
+                return random_color_part() . random_color_part() . random_color_part();
+            }
+            $catdetails = $DB->get_records_sql("SELECT qc.id, COUNT(qs.id) as qnum,
+                qc.name FROM {quiz_slots} qs, {question} q, {question_categories} qc, {question_bank_entries} qbe
+                WHERE q.id = qs.id AND qbe.questioncategoryid = qc.id AND
+                qs.quizid = ? AND q.qtype != ? GROUP BY qc.id", array($quizid, 'description'));
+            $catname = array();
```

Comment:
47 to 53..why all these arrays.. define multiple value blank arrays on a single line